File: apps/mobile/iosApp/iosApp/AssistantScreen.swift
import SwiftUI
import Shared

/// Observes the shared `AssistantViewModel`'s `StateFlow` (mirrors `HouseObservable`) and
/// owns the (untested, data/UI-layer) `da-ask` network call via `WorkerBackend.shared
/// .assistantRepository`, mirroring how the other Observables keep writes outside the pure
/// ViewModel.
@MainActor
final class AssistantObservable: ObservableObject {
    private let vm: AssistantViewModel
    @Published private(set) var state: AssistantUiState
    private var task: Task<Void, Never>?

    init(vm: AssistantViewModel = AssistantViewModel()) {
        self.vm = vm
        self.state = vm.uiState.value
        task = Task { [weak self] in
            guard let self else { return }
            for await s in self.vm.uiState { self.state = s }
        }
    }

    /// Enters the loading state immediately (optimistic, mirrors the Android host), then
    /// streams `da-ask`'s SSE response and feeds each event to the ViewModel as it
    /// arrives. Ignores blank input and re-entrancy while a request is in flight (the
    /// ViewModel itself guards this too).
    func ask(_ text: String) {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, !state.loading else { return }
        vm.onUserSubmitted(text: trimmed)
        vm.onStreamStart()
        Task {
            do {
                for try await event in WorkerBackend.shared.assistantRepository.askStream(question: trimmed) {
                    switch event {
                    case let meta as AssistantStreamEvent.Meta:
                        self.vm.onStreamMeta(
                            citations: meta.citations, deferred: meta.deferred, route: meta.route,
                            lifeSafety: meta.lifeSafety)
                    case let delta as AssistantStreamEvent.Delta:
                        self.vm.onStreamDelta(text: delta.text)
                    case let retract as AssistantStreamEvent.Retract:
                        self.vm.onStreamRetract(content: retract.content)
                    case is AssistantStreamEvent.Done:
                        self.vm.onStreamDone()
                    case let failed as AssistantStreamEvent.Failed:
                        self.vm.onError(message: failed.message)
                    default:
                        break
                    }
                }
            } catch {
                self.vm.onError(message: "Couldn't reach Snoopy. Try again.")
            }
        }
    }

    deinit { task?.cancel() }
}

/// Desk Assistant chat (V1_SCOPE §4). An empty thread shows a short intro + starter-prompt
/// chips (`AssistantPrompts.shared.starters`, shared with Android); a live thread renders
/// left/right bubbles with citation chips, a life-safety banner, and an escalation tag when
/// the answer routed to a duty contact.
struct AssistantTabView: View {
    @ObservedObject var model: AssistantObservable
    /// The Assistant has no bottom-bar item of its own — it opens from the My-Shifts FAB or
    /// the More sheet, from whichever tab the worker was on — so closing it needs somewhere
    /// to return to that isn't hardcoded to My Shifts.
    let onBack: () -> Void
    @State private var input: String = ""
    @Environment(\.colorScheme) private var scheme
    private var c: ShiftColors { .resolve(scheme) }

    var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 6) {
                Button(action: onBack) {
                    Image(systemName: ShiftIcons.chevronLeft)
                        .font(.system(size: 15, weight: .semibold))
                        .foregroundColor(c.sec)
                        .frame(width: 30, height: 30)
                        .background(c.surfaceVar)
                        .clipShape(Circle())
                }
                .buttonStyle(.plain)
                .accessibilityIdentifier("assistant_back")
                Text("Ask Snoopy")
                    .font(ShiftFont.sans(26, .bold, relativeTo: .largeTitle))
                    .foregroundColor(c.ink)
                Spacer(minLength: 0)
            }
            .padding(.leading, 10).padding(.trailing, 16).padding(.top, 10).padding(.bottom, 8)
            if model.state.messages.isEmpty {
                emptyState
            } else {
                ScrollViewReader { proxy in
                    ScrollView {
                        LazyVStack(alignment: .leading, spacing: 12) {
                            ForEach(model.state.messages, id: \.id) { message in
                                let isStreamingPlaceholder =
                                    model.state.loading && message.content.isEmpty
                                        && message.id == model.state.messages.last?.id
                                AssistantBubbleView(message: message, showShimmer: isStreamingPlaceholder)
                                    .id(message.id)
                            }
                        }
                        .padding(.horizontal, Spacing.screen)
                        .padding(.vertical, Spacing.l)
                    }
                    .onChange(of: model.state.messages.count) { _ in scrollToBottom(proxy) }
                    .onChange(of: model.state.loading) { _ in scrollToBottom(proxy) }
                    .onChange(of: model.state.messages.last?.content) { _ in scrollToBottom(proxy) }
                }
            }
            if let error = model.state.error {
                Text(error)
                    .font(ShiftFont.sans(12.5))
                    .foregroundColor(c.danger.accent)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.horizontal, Spacing.screen)
                    .padding(.top, 4)
            }
            inputBar
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(c.bg)
        // A non-wrapping marker, not the container itself — an identifier set directly on a
        // wrapping container leaks onto every descendant element in the XCUITest tree,
        // shadowing that container's own more-specific descendant identifiers (confirmed
        // empirically; see ContentView.swift's `shifts_screen` fix for the full explanation).
        .overlay(alignment: .topLeading) {
            Color.clear.frame(width: 1, height: 1).accessibilityIdentifier("assistant_screen")
        }
    }

    private func scrollToBottom(_ proxy: ScrollViewProxy) {
        guard let target = model.state.messages.last?.id else { return }
        withAnimation { proxy.scrollTo(target, anchor: .bottom) }
    }

    private var emptyState: some View {
        ScrollView {
        VStack(spacing: 0) {
            Spacer(minLength: 24)
            ZStack {
                RoundedRectangle(cornerRadius: 14, style: .continuous).fill(c.today)
                    .frame(width: 48, height: 48)
                Image(systemName: ShiftIcons.sparkles).font(.system(size: 22)).foregroundColor(c.blue)
            }
            Text("Ask a desk question")
                .font(ShiftFont.sans(19, .semibold)).foregroundColor(c.ink)
                .padding(.top, 16)
            Text("Answers are grounded in the official documentation and current duty schedule.")
                .font(ShiftFont.sans(14)).foregroundColor(c.sec)
                .multilineTextAlignment(.center)
                .padding(.top, 4).padding(.horizontal, 24)
            VStack(spacing: 8) {
                ForEach(AssistantPrompts.shared.starters, id: \.self) { prompt in
                    Button(action: { model.ask(prompt) }) {
                        Text(prompt)
                            .font(ShiftFont.sans(13.5)).foregroundColor(c.ink)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .padding(.horizontal, 14).padding(.vertical, 10)
                            .background(c.surface)
                            .clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
                            .overlay(RoundedRectangle(cornerRadius: 14, style: .continuous).strokeBorder(c.divider, lineWidth: 1))
                    }
                    .buttonStyle(.plain)
                }
            }
            .padding(.top, 20).padding(.horizontal, 24)
            Spacer(minLength: 24)
        }
        .frame(maxWidth: .infinity)
        }
    }

    private var inputBar: some View {
        HStack(spacing: 10) {
            TextField("Ask a desk question...", text: $input)
                .font(ShiftFont.sans(14.5))
                .foregroundColor(c.ink)
                .padding(.horizontal, 16).padding(.vertical, 12)
                .background(c.surfaceVar)
                .clipShape(RoundedRectangle(cornerRadius: Radii.pill, style: .continuous))
                .overlay(RoundedRectangle(cornerRadius: Radii.pill, style: .continuous).strokeBorder(c.divider, lineWidth: 1))
                .accessibilityIdentifier("assistant_input")

            let canSend = !input.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && !model.state.loading
            Button(action: send) {
                Image(systemName: ShiftIcons.send)
                    .font(.system(size: 16, weight: .semibold))
                    .foregroundColor(canSend ? .white : c.ter)
                    .frame(width: 44, height: 44)
                    .background(canSend ? c.blue : c.surfaceVar)
                    .clipShape(Circle())
            }
            .buttonStyle(.plain)
            .disabled(!canSend)
            .accessibilityIdentifier("assistant_send")
        }
        .padding(.horizontal, Spacing.screen)
        .padding(.vertical, Spacing.l)
    }

    private func send() {
        let text = input
        input = ""
        model.ask(text)
    }
}

/// One line of a parsed answer. Spans are concatenated into a single `Text` so the line still
/// wraps and hyphenates as one paragraph rather than breaking at every style change.
private struct MarkdownLineView: View {
    let line: MarkdownLine
    let isUser: Bool
    @Environment(\.colorScheme) private var scheme
    private var c: ShiftColors { .resolve(scheme) }

    var body: some View {
        let body = line.spans.reduce(Text("")) { acc, span in acc + styled(span) }
        if line.bullet {
            HStack(alignment: .top, spacing: 6) {
                Text("•").font(ShiftFont.sans(14.5)).foregroundColor(isUser ? .white : c.sec)
                body
            }
        } else {
            body.frame(maxWidth: .infinity, alignment: .leading)
        }
    }

    private func styled(_ span: MarkdownSpan) -> Text {
        var t = Text(span.text)
        if span.code {
            t = t.font(ShiftFont.mono(13.5))
        } else {
            t = t.font(ShiftFont.sans(14.5, span.bold ? .semibold : .regular))
        }
        if span.italic { t = t.italic() }
        return t.foregroundColor(isUser ? .white : c.ink)
    }
}

/// Where an answer came from, kept out of the way (BSpec §17.3).
///
/// Collapsed it is a single quiet row ("2 sources"); tapping reveals the document names, each
/// clipped to one line. The names are long binder titles, so showing them inline pushed the
/// answer off screen and clipped mid-word. The answer is what the worker needs in the moment;
/// provenance is there to be checked, not read.
private struct SourcesBar: View {
    let citations: [Citation]
    @State private var expanded = false
    @Environment(\.colorScheme) private var scheme
    private var c: ShiftColors { .resolve(scheme) }

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Button {
                withAnimation(.easeInOut(duration: 0.15)) { expanded.toggle() }
            } label: {
                HStack(spacing: 4) {
                    Image(systemName: ShiftIcons.chevronRight)
                        .font(.system(size: 9, weight: .semibold))
                        .rotationEffect(.degrees(expanded ? 90 : 0))
                    Text(citations.count == 1 ? "1 source" : "\(citations.count) sources")
                        .font(ShiftFont.sans(11.5))
                }
                .foregroundColor(c.ter)
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("assistant_sources_toggle")

            if expanded {
                VStack(alignment: .leading, spacing: 2) {
                    ForEach(Array(citations.enumerated()), id: \.offset) { _, citation in
                        Text(citation.sourceRef)
                            .font(ShiftFont.sans(11))
                            .foregroundColor(c.ter)
                            .lineLimit(1)
                            .truncationMode(.tail)
                    }
                }
                .accessibilityIdentifier("assistant_sources_list")
            }
        }
        .padding(.top, 2)
    }
}

private struct AssistantBubbleView: View {
    let message: AssistantMessage
    var showShimmer: Bool = false
    @Environment(\.colorScheme) private var scheme
    private var c: ShiftColors { .resolve(scheme) }

    var body: some View {
        let isUser = message.role == .user
        HStack {
            if isUser { Spacer(minLength: 40) }
            VStack(alignment: isUser ? .trailing : .leading, spacing: 6) {
                if !isUser && message.showSafetyBanner {
                    HStack(spacing: 8) {
                        Image(systemName: ShiftIcons.warning).font(.system(size: 14)).foregroundColor(c.danger.accent)
                        Text(message.lifeSafety ?? "")
                            .font(ShiftFont.sans(12.5, .semibold)).foregroundColor(c.danger.deep)
                    }
                    .padding(.horizontal, 12).padding(.vertical, 8)
                    .background(c.danger.tint)
                    .clipShape(RoundedRectangle(cornerRadius: 10, style: .continuous))
                }
                Group {
                    if showShimmer {
                        VStack(alignment: .leading, spacing: 6) {
                            ShimmerBar().frame(width: 160, height: 12)
                            ShimmerBar().frame(width: 110, height: 12)
                        }
                    } else {
                        // Rendered, not raw: the model writes Markdown regardless of the prompt,
                        // and this bubble used to print `**10:00 pm**` verbatim.
                        VStack(alignment: .leading, spacing: 3) {
                            ForEach(
                                Array(AssistantMarkdownKt.parseAssistantMarkdown(raw: message.content).enumerated()),
                                id: \.offset
                            ) { _, line in
                                MarkdownLineView(line: line, isUser: isUser)
                            }
                        }
                    }
                }
                    .padding(.horizontal, 14).padding(.vertical, 11)
                    .background(isUser ? c.blue : c.surface)
                    .clipShape(RoundedRectangle(cornerRadius: 16, style: .continuous))
                    .overlay {
                        if !isUser {
                            RoundedRectangle(cornerRadius: 16, style: .continuous).strokeBorder(c.divider, lineWidth: 1)
                        }
                    }
                if !isUser && !message.citations.isEmpty {
                    SourcesBar(citations: message.citations)
                }
                if let route = message.route {
                    Text("Routed to: \(route.tierLabel ?? route.resolvedTier)")
                        .font(ShiftFont.sans(12, .medium)).foregroundColor(c.sec)
                }
            }
            .frame(maxWidth: 300, alignment: isUser ? .trailing : .leading)
            if !isUser { Spacer(minLength: 40) }
        }
    }
}

File: apps/mobile/androidApp/src/main/java/com/pennhousing/shift/ui/AssistantScreen.kt
package com.pennhousing.shift.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.pennhousing.shift.shared.assistant.AssistantMessage
import com.pennhousing.shift.shared.assistant.AssistantPrompts
import com.pennhousing.shift.shared.assistant.AssistantRole
import com.pennhousing.shift.shared.assistant.AssistantStreamEvent
import com.pennhousing.shift.shared.assistant.Citation
import com.pennhousing.shift.shared.assistant.parseAssistantMarkdown
import com.pennhousing.shift.shared.data.AssistantRepository
import com.pennhousing.shift.shared.data.WorkerBackend
import com.pennhousing.shift.shared.viewmodel.AssistantViewModel
import com.pennhousing.shift.ui.kit.ShiftIcons
import com.pennhousing.shift.ui.kit.shimmer
import com.pennhousing.shift.ui.theme.ShiftTheme
import kotlinx.coroutines.launch

/**
 * Desk Assistant chat (V1_SCOPE §4). Native Compose UI over the shared, tested
 * [AssistantViewModel] — this file owns the (untested, data/UI-layer) network call via
 * [AssistantRepository], mirroring how the Shifts screen's host keeps writes outside the
 * pure ViewModel. An empty thread shows a short intro + [AssistantPrompts.starters] chips;
 * a live thread renders left/right bubbles with citation chips, a life-safety banner, and
 * an escalation tag when the answer routed to a duty contact.
 */
@Composable
fun AssistantScreen(
    vm: AssistantViewModel,
    onBack: () -> Unit,
    repository: AssistantRepository = WorkerBackend.assistantRepository,
) {
    val c = ShiftTheme.colors
    val state by vm.uiState.collectAsStateWithLifecycle()
    val scope = rememberCoroutineScope()
    var input by remember { mutableStateOf("") }
    val listState = rememberLazyListState()

    fun submit(text: String) {
        val trimmed = text.trim()
        if (trimmed.isEmpty() || state.loading) return
        vm.onUserSubmitted(trimmed)
        input = ""
        vm.onStreamStart()
        scope.launch {
            runCatching {
                repository.askStream(trimmed).collect { event ->
                    when (event) {
                        is AssistantStreamEvent.Meta ->
                            vm.onStreamMeta(event.citations, event.deferred, event.route, event.lifeSafety)
                        is AssistantStreamEvent.Delta -> vm.onStreamDelta(event.text)
                        is AssistantStreamEvent.Retract -> vm.onStreamRetract(event.content)
                        is AssistantStreamEvent.Done -> vm.onStreamDone()
                        is AssistantStreamEvent.Failed -> vm.onError(event.message)
                    }
                }
            }.onFailure { vm.onError(it.message ?: "Couldn't reach Snoopy. Try again.") }
        }
    }

    LaunchedEffect(state.messages.size, state.loading, state.messages.lastOrNull()?.content) {
        val lastIndex = state.messages.size - 1
        if (lastIndex >= 0) listState.animateScrollToItem(lastIndex)
    }

    Column(Modifier.fillMaxSize().background(c.bg)) {
        Row(
            Modifier.fillMaxWidth().padding(start = 10.dp, end = 16.dp, top = 10.dp, bottom = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            Box(
                Modifier
                    .size(30.dp)
                    .clip(RoundedCornerShape(50))
                    .background(c.surfaceVar)
                    .clickable(onClick = onBack)
                    .testTag("assistant_back"),
                contentAlignment = Alignment.Center,
            ) {
                Icon(ShiftIcons.ChevronLeft, contentDescription = "Back", tint = c.sec, modifier = Modifier.size(16.dp))
            }
            Text("Ask Snoopy", color = c.ink, fontSize = 26.sp, fontWeight = FontWeight.Bold)
        }
        Box(Modifier.weight(1f).fillMaxWidth()) {
            if (state.isEmpty) {
                AssistantEmptyState(onPrompt = ::submit)
            } else {
                LazyColumn(
                    state = listState,
                    modifier = Modifier.fillMaxSize().testTag("assistant_message_list"),
                    contentPadding = PaddingValues(horizontal = 16.dp, vertical = 12.dp),
                    verticalArrangement = Arrangement.spacedBy(12.dp),
                ) {
                    items(state.messages, key = { it.id }) { message ->
                        val isStreamingPlaceholder =
                            state.loading && message.content.isEmpty() && message.id == state.messages.last().id
                        AssistantBubble(message, showShimmer = isStreamingPlaceholder)
                    }
                }
            }
        }
        state.error?.let {
            Text(
                it,
                color = c.danger.accent,
                fontSize = 12.5.sp,
                modifier = Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 4.dp),
            )
        }
        AssistantInputBar(
            value = input,
            onValueChange = { input = it },
            onSend = { submit(input) },
            enabled = !state.loading,
        )
    }
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun AssistantEmptyState(onPrompt: (String) -> Unit) {
    val c = ShiftTheme.colors
    Column(
        Modifier.fillMaxSize().verticalScroll(rememberScrollState()).padding(horizontal = 24.dp, vertical = 24.dp),
        verticalArrangement = Arrangement.Center,
    ) {
        Box(
            Modifier.size(48.dp).clip(RoundedCornerShape(14.dp)).background(c.today),
            contentAlignment = Alignment.Center,
        ) {
            Icon(ShiftIcons.Sparkle, contentDescription = null, tint = MaterialTheme.colorScheme.primary, modifier = Modifier.size(24.dp))
        }
        Text(
            "Ask a desk question",
            modifier = Modifier.padding(top = 16.dp),
            color = c.ink,
            fontSize = 19.sp,
            fontWeight = FontWeight.SemiBold,
        )
        Text(
            "Answers are grounded in the official documentation and current duty schedule.",
            modifier = Modifier.padding(top = 4.dp),
            color = c.sec,
            fontSize = 14.sp,
            lineHeight = 19.sp,
        )
        FlowRow(
            modifier = Modifier.padding(top = 20.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            AssistantPrompts.starters.forEach { prompt ->
                Text(
                    prompt,
                    modifier =
                        Modifier
                            .clip(RoundedCornerShape(999.dp))
                            .background(c.surface)
                            .border(1.dp, c.divider, RoundedCornerShape(999.dp))
                            .clickable { onPrompt(prompt) }
                            .padding(horizontal = 14.dp, vertical = 10.dp)
                            .testTag("assistant_starter_prompt"),
                    color = c.ink,
                    fontSize = 13.5.sp,
                )
            }
        }
    }
}

@Composable
private fun AssistantBubble(
    message: AssistantMessage,
    showShimmer: Boolean = false,
) {
    val c = ShiftTheme.colors
    val isUser = message.role == AssistantRole.USER
    Row(
        Modifier.fillMaxWidth().testTag("assistant_bubble_${message.id}"),
        horizontalArrangement = if (isUser) Arrangement.End else Arrangement.Start,
    ) {
        Column(
            Modifier.widthIn(max = 300.dp),
            horizontalAlignment = if (isUser) Alignment.End else Alignment.Start,
        ) {
            if (!isUser && message.showSafetyBanner) {
                Row(
                    Modifier
                        .padding(bottom = 6.dp)
                        .clip(RoundedCornerShape(10.dp))
                        .background(c.danger.tint)
                        .padding(horizontal = 12.dp, vertical = 8.dp),
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Icon(ShiftIcons.Warning, contentDescription = null, tint = c.danger.accent, modifier = Modifier.size(16.dp))
                    Text(message.lifeSafety.orEmpty(), color = c.danger.deep, fontSize = 12.5.sp, fontWeight = FontWeight.SemiBold)
                }
            }
            Box(
                Modifier
                    .clip(RoundedCornerShape(16.dp))
                    .background(if (isUser) MaterialTheme.colorScheme.primary else c.surface)
                    .let { if (isUser) it else it.border(1.dp, c.divider, RoundedCornerShape(16.dp)) }
                    .padding(horizontal = 14.dp, vertical = 11.dp),
            ) {
                if (showShimmer) {
                    Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                        Box(Modifier.width(160.dp).height(12.dp).shimmer())
                        Box(Modifier.width(110.dp).height(12.dp).shimmer())
                    }
                } else {
                    MarkdownAnswer(
                        content = message.content,
                        color = if (isUser) MaterialTheme.colorScheme.onPrimary else c.ink,
                    )
                }
            }
            if (!isUser && message.citations.isNotEmpty()) {
                SourcesBar(message.citations)
            }
            message.route?.let { route ->
                Text(
                    "Routed to: ${route.tierLabel ?: route.resolvedTier}",
                    modifier = Modifier.padding(top = 6.dp),
                    color = c.sec,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Medium,
                )
            }
        }
    }
}

/**
 * Render an answer's Markdown instead of printing its source. Parsing is the shared
 * [parseAssistantMarkdown] so this and the SwiftUI bubble cannot drift.
 */
@Composable
private fun MarkdownAnswer(
    content: String,
    color: Color,
) {
    val lines = remember(content) { parseAssistantMarkdown(content) }
    Column(verticalArrangement = Arrangement.spacedBy(3.dp)) {
        lines.forEach { line ->
            // Spans join into ONE AnnotatedString so the line still wraps as a single
            // paragraph rather than breaking at every style change.
            val annotated =
                buildAnnotatedString {
                    line.spans.forEach { span ->
                        withStyle(
                            SpanStyle(
                                fontWeight = if (span.bold) FontWeight.SemiBold else null,
                                fontStyle = if (span.italic) FontStyle.Italic else null,
                                fontFamily = if (span.code) FontFamily.Monospace else null,
                            ),
                        ) {
                            append(span.text)
                        }
                    }
                }
            if (line.bullet) {
                Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                    Text("•", color = color, fontSize = 14.5.sp, lineHeight = 20.sp)
                    Text(annotated, color = color, fontSize = 14.5.sp, lineHeight = 20.sp)
                }
            } else {
                Text(annotated, color = color, fontSize = 14.5.sp, lineHeight = 20.sp)
            }
        }
    }
}

/**
 * Where an answer came from, kept out of the way (BSpec §17.3). Collapsed it is one quiet row;
 * tapping reveals the document names, each clipped to a single line. Binder titles are long, so
 * rendering them inline pushed the answer itself off screen.
 */
@Composable
private fun SourcesBar(citations: List<Citation>) {
    val c = ShiftTheme.colors
    var expanded by remember { mutableStateOf(false) }
    Column(Modifier.padding(top = 6.dp)) {
        Row(
            Modifier
                .clip(RoundedCornerShape(6.dp))
                .clickable { expanded = !expanded }
                .testTag("assistant_sources_toggle")
                .padding(vertical = 2.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            Icon(
                ShiftIcons.ChevronRight,
                contentDescription = null,
                tint = c.ter,
                modifier = Modifier.size(12.dp).rotate(if (expanded) 90f else 0f),
            )
            Text(
                if (citations.size == 1) "1 source" else "${citations.size} sources",
                color = c.ter,
                fontSize = 11.5.sp,
            )
        }
        if (expanded) {
            Column(
                Modifier.padding(top = 2.dp).testTag("assistant_sources_list"),
                verticalArrangement = Arrangement.spacedBy(2.dp),
            ) {
                citations.forEach { citation ->
                    Text(
                        citation.sourceRef,
                        color = c.ter,
                        fontSize = 11.sp,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
            }
        }
    }
}

@Composable
private fun AssistantInputBar(
    value: String,
    onValueChange: (String) -> Unit,
    onSend: () -> Unit,
    enabled: Boolean,
) {
    val c = ShiftTheme.colors
    Row(
        Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Row(
            Modifier
                .weight(1f)
                .clip(RoundedCornerShape(999.dp))
                .background(c.surfaceVar)
                .border(1.dp, c.divider, RoundedCornerShape(999.dp))
                .padding(horizontal = 16.dp, vertical = 12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Box(Modifier.weight(1f)) {
                if (value.isEmpty()) {
                    Text("Ask a desk question...", color = c.ter, fontSize = 14.5.sp)
                }
                BasicTextField(
                    value = value,
                    onValueChange = onValueChange,
                    modifier = Modifier.fillMaxWidth().testTag("assistant_input"),
                    singleLine = true,
                    textStyle = TextStyle(color = c.ink, fontSize = 14.5.sp),
                    cursorBrush = SolidColor(MaterialTheme.colorScheme.primary),
                )
            }
        }
        val canSend = enabled && value.isNotBlank()
        Box(
            Modifier
                .size(44.dp)
                .clip(RoundedCornerShape(999.dp))
                .background(if (canSend) MaterialTheme.colorScheme.primary else c.surfaceVar)
                .let { if (canSend) it.clickable(onClick = onSend) else it }
                .testTag("assistant_send"),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                ShiftIcons.Send,
                contentDescription = "Send",
                tint = if (canSend) MaterialTheme.colorScheme.onPrimary else c.ter,
                modifier = Modifier.size(18.dp),
            )
        }
    }
}
